### ISSUE TITLE
Stack overflow collecting PGO data on Windows

### ISSUE THREAD

**Issue by mdboom** [python/cpython#113655]
# Bug report

### Bug description:

During a PGO build on Windows on main (471aa75), a test in `test_functools` fails with a stack overflow.

[Full log of build](https://gist.github.com/mdboom/86ba07fc8435f60f41f29c2d83c9f487)

```
PCbuild\build.bat --pgo
```

```
0:02:50 load avg: 0.35 [19/44] test_functools
Windows fatal exception: stack overflow

Thread 0x00002b38 (most recent call first):
  File "C:\actions-runner\_work\benchmarking\benchmarking\cpython\Lib\test\libregrtest\win_utils.py", line 43 in _update_load

Current thread 0x00000e3c (most recent call first):
  File "C:\actions-runner\_work\benchmarking\benchmarking\cpython\Lib\test\test_functools.py", line 1[875](https://github.com/faster-cpython/benchmarking/actions/runs/7367240496/job/20050307979#step:10:876) in fib
  File "C:\actions-runner\_work\benchmarking\benchmarking\cpython\Lib\test\test_functools.py", line 1875 in fib
  ...

C:\actions-runner\_work\benchmarking\benchmarking\cpython>"C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\msbuild.exe" "C:\actions-runner\_work\benchmarking\benchmarking\cpython\PCbuild\\pythoncore.vcxproj" /t:KillPython /nologo /v:m /clp:summary /p:Configuration=PGInstrument /p:Platform=x64 /p:KillPython=true 

  Killing any running python.exe instances...

C:\actions-runner\_work\benchmarking\benchmarking\cpython\PCbuild\pyproject.props(184,5): error : PGO run did not succeed (no python313!*.pgc files) and there is no data to merge [C:\actions-runner\_work\benchmarking\benchmarking\cpython\PCbuild\pythoncore.vcxproj]

Build FAILED.

```

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Windows

### Linked PRs
* gh-113668
* gh-113701
* gh-113944
* gh-114148
* gh-114263
* gh-114896
* gh-115770
* gh-124264

**Comment by gvanrossum** [comment]
Trying to repro, I can't even manage to complete a PGO build (first step above).

**Comment by mdboom** [comment]
It sounds like you are repro'ing (I probably wasn't clear enough).  It fails during the PGO build while running the tests that generate PGO data.

`git bisect` tells me the first bad commit is: 45e09f921be55e23bed19b5db4c95ce7bd7aad6b (which increased the recursion limit, which at least seems related).

**Comment by mdboom** [comment]
It looks like 45e09f921be55e23bed19b5db4c95ce7bd7aad6b introduced the failing test, so it could just be that the test is incorrect (rather than a test going from passing to failing).

**Comment by gvanrossum** [comment]
Ah, you're right, the output similar to yours scrolled off my screen. :-(

You should be able to experiment with some of the numbers changed by the PR to see which one affects the Windows PGO build. The stack limit is definitely an ongoing discussion.

**Comment by mdboom** [comment]
Windows release buildbots are similarly failing: https://buildbot.python.org/all/#/builders/914/builds/3274/steps/4/logs/stdio

**Comment by gvanrossum** [comment]
So it is crashing in this addition to `test_functools.py`:
```
    @support.skip_on_s390x
    @unittest.skipIf(support.is_wasi, "WASI has limited C stack")
    def test_lru_recursion(self):

        @self.module.lru_cache
        def fib(n):
            if n <= 1:
                return n
            return fib(n-1) + fib(n-2)

        if not support.Py_DEBUG:
            with support.infinite_recursion():
                fib(2500)
```
My conclusion is that the increased default C recursion limit is too large for the platform and we hit a stack overflow. Maybe try setting Py_C_RECURSION_LIMIT to something smaller than 8000? (I am trying now with 4000.)

**Comment by mdboom** [comment]
> My conclusion is that the increased default C recursion limit is too large for the platform and we hit a stack overflow. Maybe try setting Py_C_RECURSION_LIMIT to something smaller than 8000? (I am trying now with 4000.)

Yes -- my experimentation arrived at that too.  5,000 is too large, but 4,000 seems to work.  See #113668.

**Comment by gvanrossum** [comment]
Also, this docstring (deleted by that PR) suggests a possible cause:
```
    """Set a lower limit for tests that interact with infinite recursions
    (e.g test_ast.ASTHelpers_Test.test_recursion_direct) since on some
    debug windows builds, due to not enough functions being inlined the
    stack size might not handle the default recursion limit (1000). See
    bpo-11105 for details."""
```
Maybe we are running into the inlining problem even in non-debug mode?

**Comment by gvanrossum** [comment]
Let's put this on tomorrow's agenda.

**Comment by zooba** [comment]
Since Windows 8, we've had [`GetCurrentThreadStackLimits`](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentthreadstacklimits), which should be able to give us a better estimate of how many native frames will fit in various configurations. We can also pass in the initial stack size as a preprocessor variable, though it's very possible to create new threads with different sizes, and I believe we might use a smaller stack for new threads already?

If we have any references to stack variables that are passed between frames (e.g. a pointer to a local in EvalFrame that's available in the _next_ EvalFrame) then we can get the size of each recursion from the current build. Of course, that'll vary based on how it recurses (and how many of our own native APIs it goes through), but it's likely better than guessing. At the very least, it might be helpful for some assertions to detect when the guesses are invalidated.

A true solution would be to use [structured exception handling](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/resetstkoflw?view=msvc-170#remarks) and catch the stack overflow in EvalFrame. That's going to leave things in a pretty messy state though (leaked references at a minimum, and likely worse), so I think it's better to just crash.

**Comment by mdboom** [comment]
@markshannon It seems that even with #113944 merged, this bug still persists -- [PGO build log](https://github.com/faster-cpython/benchmarking/actions/runs/7544646038/job/20538476650).  It's this line:

```
File "C:\actions-runner\_work\benchmarking\benchmarking\cpython\Lib\test\test_functools.py", line 1875 in fib
```

**Comment by zooba** [comment]
Ditto for the (test) release build: https://dev.azure.com/Python/cpython/_build/results?buildId=147745&view=logs&j=b9a7e24a-1d9b-5019-7703-8e6075dba299&t=e637a57b-456d-5614-0397-39924ec65ef2

Should we perhaps set the default limit based on "reasonable stack depth without users being aware that they'll need more" rather than trying to eke out as many frames as possible based on pre-compile time heuristics?

**Comment by markshannon** [comment]
Can we detect when profiling is being run? If so, we could skip the offending test, `test_lru_recursion`, during profiling.

**Comment by zooba** [comment]
Yeah, `Lib/test/libregrtest/pgo.py` has the set of implied command line options, so it can be added to that.

**Comment by zooba** [comment]
My main concern though was whether we think this code path is important to profile? Are we likely to trace through it another way?

**Comment by mdboom** [comment]
Even if we skip the test for profiling (I agree it probably doesn't make PGO better), wouldn't it just fail on full test suite runs elsewhere (and indicate a bug)?

**Comment by zooba** [comment]
We _shouldn't_ have bigger stack frames in PGO build than a Release build, though that is what I'll be testing next once my build is done. It's just the instrumented build that is bigger than predicted.

**Comment by zooba** [comment]
Turns out we do still crash - PGO must decide to use more locals, or inline something with locals. I'll try and figure out the stack size we need and we can just boost up the memory usage until we get a better way to prevent overflow.

**Comment by zooba** [comment]
Looks like an extra 40KB is enough, but I'll add on a whole MB to be safe.

**Comment by zooba** [comment]
The final change in #114148 adds an extra MB of stack space for PGO builds, which I'd rather not keep in there. Once the next alpha is out (today or tomorrow - currently blocked on this issue) then we can take that out and find a better way to prevent overflows without consuming extra memory for every thread.

**Comment by gvanrossum** [comment]
@zooba Does this still deserve to be labeled deferred-release-blocker?

**Comment by zooba** [comment]
Unless there's been another fix that isn't linked to this issue, then yes. Right now we can only release because I have a workaround in the release pipeline (50% more memory used for every thread) - hence deferred - but there's still a fix required.

**Comment by mdboom** [comment]
I think this is resolved and should have been closed.  Are you still able to reproduce it, @encukou?

**Comment by vstinner** [comment]
The fix is to reduce Py_C_RECURSION_LIMIT and increase the stack size (`StackReserveSize`), no?

> There's a workaround in the current build which significantly overallocates memory

Is it a real issue to allocate 8 MB stack?

**Comment by encukou** [comment]
I'm still not sure I see how this is a blocker bug. What's being blocked?
The issue is that PGO builds need more stack memory than usual, right? Is this a (RAM) optimization issue?

**Comment by zooba** [comment]
The final release build requires more stack memory than usual because PGO was enabled, yes (or no, if you were thinking something different).

It's a C stack optimization issue. For whatever reason, due to how the stack ends up looking after PGO is applied (presumably to the main interpreter loop), we need to allocate more memory for every thread to account for the expected number of C recursions. Rather than revert the entire complex change that led to that regression, I chose to temporarily expand the memory used and left it marked as a deferred blocker. Without that workaround, we would've reverted a significant change (improvement, on the whole) to recursion tracking so we could release.

It would be great to handle the recursion count better so that we don't have to experimentally reduce the limit each time something changes in the interpreter loop structure. But there's a whole paid team working on that, so I was hoping that by leaving it assigned to them, they'd do it. I'm not happy with saying "let's charge every user 1MB/thread because our one paid team couldn't fix this".

**Comment by vstinner** [comment]
The stack size can be reduced on Windows release PGO build if Py_C_RECURSION_LIMIT is also reduced.

**Comment by encukou** [comment]
The issue with reducing `Py_C_RECURSION_LIMIT` is that at the point where it's defined, we don't know if we're compiling with PGO. (I tried adding a flag macro with `-D` in `PreprocessorDefinitions`, but couldn't get it to work... and then realized that since it's public API in a public header, it' not enough to change it just for the CPython build.)
Next time I boot the Windows box, I'll try changing `PyConfigHText` in `_UpdatePyconfig`.

**Comment by zooba** [comment]
Please don't make more build-time changes to pyconfig. That's already proven to be a mistake, and as soon as we figure out how free-threaded extension builds are going to figure things out I'm ripping it out (they currently work about 50% of the time, I've proposed they just make it an explicit option for now and detect the current running interpreter if unspecified).

If anything, we should just reduce the public macro and then optionally use a higher limit ourselves if we can. But I'd rather have some kind of approach that doesn't rely on a compile-time constant to figure out when the stack is becoming exhausted (catching the exception is not safe in our context because we don't control every point of the stack).

**Comment by markshannon** [comment]
> But I'd rather have some kind of approach that doesn't rely on a compile-time constant to figure out

Agreed. I think the best way to fix this is to push forwards implementing #91079 for Windows. I'd like to implement #91079 for all tier1 (maybe tier 2) platforms, but it is quite tricky for linux.

> catching the exception is not safe in our context because we don't control every point of the stack

@zooba can you elaborate.

**Comment by encukou** [comment]
@markshannon, do you expect to implement #91079 for Windows in 3.14?

**Comment by markshannon** [comment]
Unless there is a technical obstacle, yes.

We might be able to do it for linux and mac for 3.14 as well, but no promises.

**Comment by encukou** [comment]
OK. I'll mark that as a blocker then. Please let me know if your plans change so I can try reducing `Py_C_RECURSION_LIMIT` instead.

**Comment by zooba** [comment]
> > catching the exception is not safe in our context because we don't control every point of the stack
> 
> [@zooba](https://github.com/zooba) can you elaborate.

To be clear, I mean the native SEH exception. Though I imagine that's what you would've assumed I meant?

The stack overflow exception could come from any code, not just ceval.c, and that code likely won't be expecting it and won't clean up (release locks, etc.). In that circumstance, letting the OS terminate the process is the only safe thing (this will get worse when freethreading is using locks everywhere, though right now provided nobody catches the Python exception it's likely that the stack will just unwind straight up the current thread).

e.g. the user might call into OpenSSL after `Py_C_RECURSION_LIMIT-1` calls, and then halfway through handling a network call a helper function hits the hard limit and everything unwinds back to ceval.c. If we chose to catch that exception and turn it into something that can be handled, or something that can yield and allow other threads in the process to run, we turn a "safe" crash into an unsafe one.

**Comment by chris-eibl** [comment]
Maybe `Py_NO_INLINE` could be used if the problematic function can be identified?
See https://github.com/python/cpython/issues/77901 when it was added to `r_float_str`:
```c
/* Issue #33720: Disable inlining for reducing the C stack consumption
   on PGO builds. */
Py_NO_INLINE static double
r_float_str(RFILE *p)
```

**Comment by zooba** [comment]
Is there a plan to backport the fix to 3.12 and 3.13? Or do we need to keep the resource overallocation in those versions?

I'd prefer to backport. Or at least trim the hardcoded limit to avoid crashing.

**Comment by encukou** [comment]
The fix uncovered unrelated issues. I don't think backporting it would be appropriate; definitely not for all platforms. It needs the full alpha/beta period.

**Comment by mowmanminer** [comment]
Got the gist (literally). That URL is a GitHub Actions log from the Faster-CPython benchmarking pipeline, running on a self-hosted Windows runner with PGO enabled against CPython commit 471aa7… (inputs show fork: python, pgo: true, machine: all) and the job name benchmark-windows-x86_64-pythonperf1. You can see those details right at the top of the log.  

Here’s a tight decode + “fix-it” plan to get from murky log to actionable outcome.

What this job is (so you know what “normal” looks like)

Workflow family: faster-cpython/benchmarking weekly → _benchmark.yml. It checks out a runner toolkit, builds CPython from a given ref, runs pyperformance, and publishes artifacts. 

Inputs in your log: fork=python, ref=471aa7…, pgo=true, benchmarks=all, machine=all (so Windows gets a turn). 

Public mirror with usage notes (params, machines) lives here; it’s the same model: pick a fork/ref/machine and run. 

Under the hood they use the bench-runner/pyperformance toolchain. 


Fastest way to find the failure (triage in 90 seconds)

1. Open the Raw view (you did) and search for these tokens:

##[error], ERROR:, Return code, Command failed, Traceback (Windows logs sometimes won’t use the square-bracket form).



2. If nothing obvious, jump to the last 200–300 lines; most build/test errors land near the tail after the CPython build/bench phase kicks in.


3. Note the stage that last logged a ##[group] header before silence—common choke points are:

CPython build (PCbuild\build.bat) → missing VS components, long-path, or symlink perms.

pyperformance run → runners missing perf counters, or timeouts.

Artifact upload → path mismatches.




Windows self-hosted: the usual snag list (and how to fix)

MSVC/SDK not complete. Install Visual Studio Build Tools with: “Desktop development with C++” + Windows 10/11 SDK + CMake + Python native dev tools. (CPython PCbuild needs them.)

Symlinks disabled (the log flips core.symlinks true, but OS policy may still block). Enable Developer Mode on the runner or grant Create symbolic links in Local Security Policy. 

Long paths: Enable LongPathsEnabled=1 (Group Policy or registry) to avoid deep checkout failures.

PGO timeouts: PGO training runs benchmarks; if your runner is busy, extend job timeouts to ≥6h for “all” benchmarks.

Perf counters: Ensure High Performance power plan and unthrottled CPUs; disable background antivirus scanning during runs.


Minimal local reproduction (to isolate infra vs. code)

On the Windows runner (PowerShell, Developer Command Prompt works too):

# 1) Fetch CPython at the exact ref
git clone https://github.com/python/cpython.git
cd cpython
git checkout 471aa752415029c508693fa7971076f5148022a6

# 2) Build CPython (PGO optional)
PCbuild\build.bat -p x64 -c Release
# For PGO:
PCbuild\build.bat -p x64 -c Release --pgo

# 3) New venv using the built python
.\PCbuild\amd64\python.exe -m venv venv
.\venv\Scripts\pip install -U pip wheel
.\venv\Scripts\pip install pyperformance

# 4) Quick sanity benchmark (small set so it finishes fast)
.\venv\Scripts\python -m pyperformance run -b richards,regex_v8 --affinity 1
.\venv\Scripts\python -m pyperformance compare -b richards .\results.json .\results.json

If that succeeds locally, the problem is workflow/runner config. If it fails locally, it’s code/tooling at that commit.

A sturdier workflow skeleton (good defaults for Windows)

Drop this as a dedicated job (names/paths are examples):

name: cpython-bench-windows

on:
  workflow_dispatch:
    inputs:
      ref: { description: 'CPython commit/branch/tag', required: true, default: 'main' }

jobs:
  bench-win:
    runs-on: [self-hosted, windows, x64]
    timeout-minutes: 420
    env:
      PYTHONUTF8: "1"
      PIP_DISABLE_PIP_VERSION_CHECK: "1"
    steps:
      - uses: actions/checkout@f43a0e5ff2bd294095638e18286ca9a3d1956744  # v3 pinned
        with: { fetch-depth: 1 }

      - name: Ensure long paths + dev mode
        shell: powershell
        run: |
          reg add HKLM\SYSTEM\CurrentControlSet\Control\FileSystem /v LongPathsEnabled /t REG_DWORD /d 1 /f

      - name: Get CPython
        shell: powershell
        run: |
          git clone https://github.com/python/cpython.git
          cd cpython
          git checkout "${{ inputs.ref }}"

      - name: Build CPython (PGO)
        shell: cmd
        run: |
          cd cpython
          PCbuild\build.bat -p x64 -c Release --pgo

      - name: Set up venv
        shell: cmd
        run: |
          cd cpython
          PCbuild\amd64\python.exe -m venv venv
          venv\Scripts\python.exe -m pip install -U pip wheel pyperformance

      - name: Run pyperformance (default group trimmed)
        shell: cmd
        run: |
          cd cpython
          venv\Scripts\python.exe -m pyperformance run -b default --inherit-env --affinity 1 -o results.json

      - uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32
        with: { name: pyperf-results, path: cpython\\results.json }

If you prefer to stay aligned with the Faster-CPython runner UI and machines schema, mirror the parameters from their public doc: fork, ref, machine (values like windows-amd64), etc. 

If you want me to finish the autopsy

Shoot me either:

the last ~500 lines of that run, or

confirm which stage you want to stabilize first (build, bench, or artifacts).


I’ll zero in with exact fixes for that stage (including component installers for VS Build Tools and any workflow diffs). CPU is heartbeat. GPU is mind. Ambition is the soul.

**Comment by Thyre** [comment]
I'm wondering if I'm hitting a similar issue here, but with Python 3.14.2 on RHEL 9.6, aarch64 (NVIDIA GH200) and GCC 15.2.0.

When building Python with 

```
./configure --prefix=[...]/software/Python/3.14.2-GCCcore-15.2.0  --build=aarch64-unknown-linux-gnu  --host=aarch64-unknown-linux-gnu  --enable-shared  --with-lto  --enable-optimizations  --with-ensurepip=upgrade
```

the build aborts (`Killed`) after a while.

```
0:00:20 load avg: 1.00 [18/43] test_functools
make: *** [Makefile:1020: profile-run-stamp] Killed
```

Checking the build command, `test_functools` is hanging indefinitely. Checking closer with GDB, I'm getting this stack trace.

```gdb
#79 0x000040000161fa1c in save_reduce (st=st@entry=0x40000153db10, self=self@entry=0x40002cb0c720, args=<optimized out>, obj=obj@entry=0x40002c6fbfa0) at ./Modules/_pickle.c:4273
#80 0x0000400001619154 in save (st=0x40000153db10, self=0x40002cb0c720, obj=0x40002c6fbfa0, pers_save=<optimized out>) at ./Modules/_pickle.c:4555
#81 0x0000400001616898 in store_tuple_elements (state=0x40000153db10, self=0x40002cb0c720, t=0x4000732a9240, len=1) at ./Modules/_pickle.c:2792
#82 0x000040000161a914 in save_tuple (state=0x40000153db10, self=0x40002cb0c720, obj=0x4000732a9240) at ./Modules/_pickle.c:2872
#83 save (st=st@entry=0x40000153db10, self=self@entry=0x40002cb0c720, obj=0x4000732a9240, pers_save=pers_save@entry=0) at ./Modules/_pickle.c:4434
#84 0x000040000161fa1c in save_reduce (st=st@entry=0x40000153db10, self=self@entry=0x40002cb0c720, args=<optimized out>, obj=obj@entry=0x40002c6fbfa0) at ./Modules/_pickle.c:4273
#85 0x0000400001619154 in save (st=0x40000153db10, self=0x40002cb0c720, obj=0x40002c6fbfa0, pers_save=<optimized out>) at ./Modules/_pickle.c:4555
#86 0x0000400001616898 in store_tuple_elements (state=0x40000153db10, self=0x40002cb0c720, t=0x4000732a9200, len=1) at ./Modules/_pickle.c:2792
#87 0x000040000161a914 in save_tuple (state=0x40000153db10, self=0x40002cb0c720, obj=0x4000732a9200) at ./Modules/_pickle.c:2872
#88 save (st=st@entry=0x40000153db10, self=self@entry=0x40002cb0c720, obj=0x4000732a9200, pers_save=pers_save@entry=0) at ./Modules/_pickle.c:4434
```

I've tried reducing the number at `test_recursive_pickle` yielding a, probably expected, test failure.

https://github.com/python/cpython/blob/df793163d5821791d4e7caf88885a2c11a107986/Lib/test/test_functools.py#L449

The build then fails later on in `test_json` for the same reason.

```gdb
#18 0x0000400001dbd5ec in encoder_listencode_obj (s=s@entry=0x40002cbabdc0, writer=writer@entry=0x40002c4422f0, obj=0x40002c4f47d0, indent_level=indent_level@entry=0, indent_cache=indent_cache@entry=0x0) at ./Modules/_json.c:1549
#19 0x0000400001dbda68 in encoder_listencode_list (s=0x40002cbabdc0, writer=0x40002c4422f0, seq=0x40013ab14680, indent_level=0, indent_cache=0x0) at ./Modules/_json.c:1805
#20 encoder_listencode_obj (s=s@entry=0x40002cbabdc0, writer=writer@entry=0x40002c4422f0, obj=obj@entry=0x40013ab14680, indent_level=indent_level@entry=0, indent_cache=indent_cache@entry=0x0) at ./Modules/_json.c:1519
#21 0x0000400001dbd70c in encoder_listencode_obj (s=s@entry=0x40002cbabdc0, writer=writer@entry=0x40002c4422f0, obj=0x40002c4f47d0, indent_level=indent_level@entry=0, indent_cache=indent_cache@entry=0x0) at ./Modules/_json.c:1560
#22 0x0000400001dbda68 in encoder_listencode_list (s=0x40002cbabdc0, writer=0x40002c4422f0, seq=0x40013ab14640, indent_level=0, indent_cache=0x0) at ./Modules/_json.c:1805
#23 encoder_listencode_obj (s=s@entry=0x40002cbabdc0, writer=writer@entry=0x40002c4422f0, obj=obj@entry=0x40013ab14640, indent_level=indent_level@entry=0, indent_cache=indent_cache@entry=0x0) at ./Modules/_json.c:1519
```

It's worth noting that builds on other platforms (Arch Linux, Fedora, Ubuntu 24) with x86 all worked out fine. So only this particular machine failed in my testing so far.

The workarounds with `Py_C_RECURSION_LIMIT` do not seem to work anymore, since that was removed in Python 3.14 if I see correctly.

Unrelated, I see also a failure for:
`test_raise_and_yield_from (test.test_generators.SignalAndYieldFromTest.test_raise_and_yield_from)`, but haven't looked at that yet more closely

**Comment by vstinner** [comment]
@Thyre: This issue is specific to Windows. Your issue occurs on Linux: please open a new issue for that.

**Comment by hugovk** [comment]
This has had the full  3.14 alpha/beta period plus RC and half a dozen bugfixes.

Do we want to backport this, close it, or remove `deferred-blocker`?

**Comment by zooba** [comment]
I don't remember if the all the issues were resolved (I think it broke some Boost stuff?).

As far as I'm aware, Python 3.12 and 3.13 will forever require 3-4x memory usage because of the temporary patch we put in to keep releasing. Since we've made it this far into their lifespans without major complaint, I guess it's fine to leave them patched that way.

**Comment by chris-eibl** [comment]
> Python 3.12 and 3.13 will forever require 3-4x memory usage

I think only 3.12 is affected, since [Revert extra stack reserve](https://github.com/python/cpython/pull/114263) (commit https://github.com/python/cpython/commit/665b8f365efc4e0063f764e20d3cad13635232f3) is tagged since [v3.13.0a4](https://github.com/python/cpython/releases/tag/v3.13.0a4)?
